tm2/pkg/sdk/auth/keeper.go

```diff
@@ -160,3 +164,148 @@ func (ak AccountKeeper) decodeAccount(bz []byte) (acc std.Account) {
+//
+// The math formula is an abstraction of a simple solution for the underlying problem we're trying to solve.
+// 1. What do we do if the gas used is less than the target gas in a block?
+// 2. How do we bring the gas used back to the target level, if gas used is more than the target?
+// We simplify the solution with a one-line formula to explain the idea. However, in reality, we need to treat
+// two scenarios differently. For example, in the first case, we need to increase the gas by at least 1 unit,
+// instead of round down for the integer divisions, and in the second case, we should set a floor
+// as the target gas price. This is just a starting point. Down the line, the solution might not be even
+// representable by one simple formula
+
+func (gk GasPriceKeeper) calcBlockGasPrice(lastGasPrice std.GasPrice, gasUsed int64, maxGas int64, params Params) std.GasPrice {
+	// If no block gas price is set, there is no need to change the last gas price.
+	if lastGasPrice.Price.Amount == 0 {
+		return lastGasPrice
+	}
+
+	// This is also a configuration to indicate that there is no need to change the last gas price.
+	if params.TargetGasRatio == 0 {
+		return lastGasPrice
+	}
+	// if no gas used, no need to change the lastPrice
+	if gasUsed == 0 {
+		return lastGasPrice
+	}
+	var (
+		num   = new(big.Int)
+		denom = new(big.Int)
+	)
+
+	// targetGas = maxGax*TargetGasRatio/100
+
+	num.Mul(big.NewInt(maxGas), big.NewInt(params.TargetGasRatio))
+	num.Div(num, big.NewInt(int64(100)))
+	targetGasInt := new(big.Int).Set(num)
+
+	// if used gas is right on target, no need to change
+	gasUsedInt := big.NewInt(gasUsed)
+	if targetGasInt.Cmp(gasUsedInt) == 0 {
+		return lastGasPrice
+	}
+
+	c := params.GasPricesChangeCompressor
+	lastPriceInt := big.NewInt(lastGasPrice.Price.Amount)
+
+	if gasUsedInt.Cmp(targetGasInt) == 1 { // gas used is more than the target
+		// increase gas price
+		num = num.Sub(gasUsedInt, targetGasInt)
+		num.Mul(num, lastPriceInt)
+		num.Div(num, targetGasInt)
+		num.Div(num, denom.SetInt64(c))
+		// increase at least 1
+		diff := max(num, big.NewInt(1))
+		num.Add(lastPriceInt, diff)
+		// XXX should we cap it with a max gas price?
+	} else { // gas used is less than the target
+		// decrease gas price down to initial gas price
+		initPriceInt := big.NewInt(params.InitialGasPrice.Price.Amount)
+		if lastPriceInt.Cmp(initPriceInt) == -1 {
+			return params.InitialGasPrice
+		}
+		num.Sub(targetGasInt, gasUsedInt)
+		num.Mul(num, lastPriceInt)
+		num.Div(num, targetGasInt)
+		num.Div(num, denom.SetInt64(c))
+
+		num.Sub(lastPriceInt, num)
+		// gas price should not be less than the initial gas price,
+		num = max(num, initPriceInt)
+	}
+
+	if !num.IsInt64() {
+		panic("The min gas price is out of int64 range")
+	}
+
+	lastGasPrice.Price.Amount = num.Int64()
+	return lastGasPrice
+}
+
+// max returns the larger of x or y.
+func max(x, y *big.Int) *big.Int {
```

Comment:
Nitpick: maybe consider naming this `maxBig`

Reply:
makes sense, done.